Improve `Lifetime::suggestion`

nnethercote opened the pull request:
r? @lcnr

nnethercote committed:
Augment `impl-trait-missing-lifetime-gated.rs`.

We have coverage for `Foo` and `Foo<T>` but not for `Foo<>`. This commit
adds it. Note that the output has bogus syntax: `impl Foo'a, >`

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14124110906/job/39569498433) [(plain)](https://github.com/rust-lang/rust/commit/be80b190c14184404da8b52bf244961bcf8f5ca9/checks/39569498433/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#19 exporting to docker image format
#19 sending tarball 20.1s done
#19 DONE 33.5s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
   Compiling rustc_pattern_analysis v0.0.0 (/checkout/compiler/rustc_pattern_analysis)
   Compiling rustc_ast_lowering v0.0.0 (/checkout/compiler/rustc_ast_lowering)
   Compiling rustc_query_impl v0.0.0 (/checkout/compiler/rustc_query_impl)
   Compiling rustc_smir v0.0.0 (/checkout/compiler/rustc_smir)
error[E0433]: failed to resolve: use of undeclared type `IsAnonInPath`
    --> compiler/rustc_ast_lowering/src/lib.rs:1807:38
     |
1807 |         if matches!(is_anon_in_path, IsAnonInPath::Yes { .. }) {
     |                                      ^^^^^^^^^^^^ use of undeclared type `IsAnonInPath`
     |
help: consider importing one of these type aliases
     |
45   + use crate::hir::IsAnonInPath;
     |
45   + use rustc_hir::IsAnonInPath;
     |

For more information about this error, try `rustc --explain E0433`.
[RUSTC-TIMING] rustc_ast_lowering test:false 1.457
error: could not compile `rustc_ast_lowering` (lib) due to 1 previous error
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14124372283/job/39570246033) [(plain)](https://github.com/rust-lang/rust/commit/effd5cca67bc8db1fefaf524f908507c7fbc111b/checks/39570246033/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] rustc_transmute test:false 0.403
error[E0599]: no variant or associated item named `No` found for enum `Option` in the current scope
  --> compiler/rustc_hir/src/hir/tests.rs:61:52
   |
61 |                     is_anon_in_path: IsAnonInPath::No,
   |                                                    ^^ variant or associated item not found in `Option<AngleBrackets>`

    Checking rustc_infer v0.0.0 (/checkout/compiler/rustc_infer)
[RUSTC-TIMING] rustc_next_trait_solver test:true 1.364
    Checking rustc_pattern_analysis v0.0.0 (/checkout/compiler/rustc_pattern_analysis)
For more information about this error, try `rustc --explain E0599`.
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #139826) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote committed:
Handle `Path<>` better in error messages.

`Path<>` needs to be distinguished from `Path<T>`. This commit does
that, improving some error messages.

nnethercote commented:
Details in individual commits.

cc @shepmaster

shepmaster commented on compiler/rustc_ast_lowering/src/path.rs:
> Also, `maybe_insert_elided_lifetimes_in_path` now does a little more
span shrinkage so all the spans it produces are empty and point to the
lifetime insertion point. As a result, `suggestion` no longer needs to
do any span shrinking.

I'm no estebank, but I disagree that this improves the error reporting. I actually specifically [go out of my way](https://github.com/rust-lang/rust/blob/8d5683060a7fa9515cc85d6037ccc26cc1aa84bf/compiler/rustc_lint/src/lifetime_style.rs#L250-L259) to avoid that kind of diagnostic span.

Specifically, I think that a suggestion that points the user to nothing (or in this case, the unrelated closing parenthesis) is very confusing:

```
 LL |     fn f(_: impl Foo) {}
    |                     ^ expected named lifetime parameter
```

Instead, I expand the location to the nearest type in these cases, broader than strictly necessary:

```diff
 LL |     fn f(_: impl Foo) {}
-   |                     ^ expected named lifetime parameter
+   |                  ^^^ expected named lifetime parameter
```

I have less of a strong opinion on the `Path<>` case as I don't think it's super common for a human to write it like that.

compiler-errors replied:
I think the span at the end of the identifier is fine, but would be better if we could tweak the note to mention that we expected a generics args list afterwards. Is that possible as a compromise?

nnethercote replied:
I've just reverted this part of the change.

lcnr commented on compiler/rustc_hir/src/hir.rs:
```diff
@@ -304,13 +310,18 @@ impl Lifetime {
-                (self.ident.span.shrink_to_hi(), format!("<{new_lifetime}>"))
+            (Hidden, Path { angle_brackets: AngleBrackets::Missing }) => {
+                (self.ident.span, format!("<{new_lifetime}>"))
             }
```
you can merge all these branches, can you not?

lcnr commented on compiler/rustc_hir/src/hir.rs:
```diff
@@ -304,13 +310,18 @@ impl Lifetime {
+
+            // The user wrote `Path<>`, and omitted the `'_`..
+            (Hidden, Path { angle_brackets: AngleBrackets::Empty }) => {
                 (self.ident.span, format!("{new_lifetime}, "))
```
could we avoid the `, ` here? and just emit `'a`?

nnethercote replied:
Good catch, it was printing `Foo<'a, >` in the suggestion. I will fix.

nnethercote commented:
Ok, I have reduced the PR to perturb the carets in the error messages less. It now only makes the changes necessary to fix the malformed error messages.

lcnr commented:
@bors r+ rollup

bors commented:
:pushpin: Commit d42edee451b87ce3d7b9de82f43c142592e1e661 has been approved by `lcnr`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=lcnr d42edee451b87ce3d7b9de82f43c142592e1e661 -->
<!-- homu: {"type":"Approved","sha":"d42edee451b87ce3d7b9de82f43c142592e1e661","approver":"lcnr","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
